### ISSUE TITLE
权重文件可以导出onnx模型，但使用onnxruntime加载推理却异常？

### ISSUE THREAD

**Issue by wkexinw** [sirius-ai/LPRNet_Pytorch#99]
有使用权重转onnx并加载推理的例子吗？
提示错误：
❌ ONNX Runtime 加载失败: [ONNXRuntimeError] : 1 : FAIL : Load model from ./weights/Final_LPRNet_model.onnx failed:Node (/original_model/backbone.3/MaxPool) Op (MaxPool) [ShapeInferenceError] Attribute dilations has incorrect size
💡 可能原因：MaxPool3d 的 dilations 属性不兼容，建议使用 Netron 查看模型结构

`# -*- coding: utf-8 -*-
# /usr/bin/env/python3

import torch
import cv2
import numpy as np
import argparse
import os
from model.LPRNet import build_lprnet, small_basic_block
from data.load_data import CHARS
import torch.nn as nn
import onnx
import onnxruntime

def preprocess_image(image_path):
    # 读取图片
    img = cv2.imread(image_path)
    # 调整大小为模型输入尺寸 (94, 24)
    img = cv2.resize(img, (94, 24))
    # BGR转RGB
    img = cv2.cvtColor(img, cv2.COLOR_BGR2RGB)
    # 归一化
    img = img.astype(np.float32) / 255.0
    # 调整维度：[H, W, C] -> [C, H, W]
    img = np.transpose(img, (2, 0, 1))
    # 添加批次维度：[C, H, W] -> [B, C, H, W]
    img = np.expand_dims(img, axis=0)
    # 添加虚拟深度维度：[B, C, H, W] -> [B, C, D, H, W]，其中D=1
    img = np.expand_dims(img, axis=2)  # 在第3个位置添加深度维度
    return img
# -------------------------------
# 2. 加载 .pth 权重并导出 ONNX
# -------------------------------

# 设置参数（根据你的训练配置调整）
lpr_max_len = 8
class_num = 68      # 修复：将类别数从66改为68，以匹配预训练权重
dropout_rate = 0.5
phase = False       # eval mode

# 创建模型实例
model = build_lprnet(lpr_max_len=lpr_max_len, phase=phase, class_num=class_num, dropout_rate=dropout_rate)

# 加载 .pth 权重（state_dict）
pth_path = "./weights/Final_LPRNet_model.pth"  # 替换为你的 .pth 文件路径
checkpoint = torch.load(pth_path, map_location='cpu')
if isinstance(checkpoint, dict) and 'state_dict' in checkpoint:
    model.load_state_dict(checkpoint['state_dict'])
else:
    model.load_state_dict(checkpoint)  # 直接是 state_dict
model.eval()

print("✅ 模型权重加载成功")

# -------------------------------
# 3. 准备 5D 输入 [B, C, D, H, W]
# 修改：添加虚拟的Depth维度D=1，以兼容MaxPool3d操作
# -------------------------------

dummy_input = torch.randn( 1,3,24, 94)  # [B, C, D, H, W] 格式，D=1是虚拟的深度维度

# -------------------------------
# 4. 导出为 ONNX
# -------------------------------
onnx_path = "./weights/Final_LPRNet_model.onnx"

# 创建一个兼容ONNX的包装模型类
class ONNXCompatibleModel(torch.nn.Module):
    def __init__(self, original_model):
        super(ONNXCompatibleModel, self).__init__()
        self.original_model = original_model
    
    def forward(self, x):
        # 确保输入是4D格式 [B, C, H, W]
        if len(x.shape) == 5:
            # 如果是5D输入，对深度维度取平均转换为4D
            x = x.mean(dim=2)
        return self.original_model(x)

# 创建兼容模型
onnx_compatible_model = ONNXCompatibleModel(model)
onnx_compatible_model.eval()

# 使用4D输入进行导出
dummy_input_4d = torch.randn(1, 3, 24, 94)  # [B, C, H, W] 格式

# 导出为ONNX
torch.onnx.export(
    model=onnx_compatible_model,
    args=dummy_input_4d,
    f=onnx_path,
    input_names=["input"],           # 输入节点名
    output_names=["output"],         # 输出节点名
    opset_version=20,                # 使用更高版本的opset
    do_constant_folding=True,        # 优化常量折叠
    verbose=False,
    # 关键参数：控制导出行为
    dynamic_axes={
        'input': {0: 'batch_size'},
        'output': {0: 'batch_size'}
    }
)

print(f"✅ ONNX 模型已成功导出: {onnx_path}")

# -------------------------------
# 5. 验证 ONNX 模型是否可被 ONNX Runtime 加载
# -------------------------------
try:
    import onnxruntime as ort

    onnx.checker.check_model(onnx_path)
    model = onnx.load(onnx_path)
    print("✅ ONNX 模型结构检查通过")
    #使用model进行推理
    ort_session = ort.InferenceSession(onnx_path)
    print("✅ ONNX Runtime 会话创建成功")
except Exception as e:
    print("❌ ONNX Runtime 加载失败:", str(e))
    print("💡 可能原因：MaxPool3d 的 dilations 属性不兼容，建议使用 Netron 查看模型结构")`

**Comment by wkexinw** [comment]
可以加载，可能虚拟环境问题。关闭。
